File: leetcode/medium/0200. Number of Islands/main.go
package main

type Position struct {
	X int
	Y int
}

var directions = []Position{
	{X: -1, Y: 0},
	{X: 0, Y: 1},
	{X: 1, Y: 0},
	{X: 0, Y: -1},
}

func numIslands(grid [][]byte) int {
	rows, cols := len(grid), len(grid[0])

	visited := make([][]bool, rows)
	for i := 0; i < rows; i++ {
		visited[i] = make([]bool, cols)
	}

	var dfs func(start Position)
	dfs = func(start Position) {
		if start.X < 0 || start.X >= rows || start.Y < 0 || start.Y >= cols || visited[start.X][start.Y] || grid[start.X][start.Y] == '0' {
			return
		}

		visited[start.X][start.Y] = true

		for _, direction := range directions {
			dfs(Position{
				X: start.X + direction.X,
				Y: start.Y + direction.Y,
			})
		}
	}

	result := 0

	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if !visited[i][j] && grid[i][j] == '1' {
				result += 1
				dfs(Position{
					X: i,
					Y: j,
				})
			}
		}
	}

	return result
}

File: leetcode/medium/0200. Number of Islands/main.java
// Solved by @kitanoyoru
// https://leetcode.com/problems/number-of-islands/

class Solution {
  int[][] DIRECTIONS = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

  private void dfs(int sr, int sc, char[][] grid, boolean[][] visited, int rows, int cols) {
    Queue<int[]> q = new LinkedList<>();

    q.add(new int[] {sr, sc});
    visited[sr][sc] = true;

    while (!q.isEmpty()) {
      int[] entry = q.remove();
      for (var d : DIRECTIONS ) {
        int newRow = entry[0] + d[0];
        int newCol = entry[1] + d[1];
        if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols && !visited[newRow][newCol] && grid[newRow][newCol] == '1') {
          q.add(new int[] {newRow, newCol});
          visited[newRow][newCol] = true;
        }
      }
    }
  }

  public int numIslands(char[][] grid) {
    int rows = grid.length, cols = grid[0].length;
    int counter = 0;
    boolean[][] visited = new boolean[rows][cols];

    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        if (grid[i][j] == '1' && !visited[i][j]) {
          dfs(i, j, grid, visited, rows, cols);
          counter++;
        }
      }
    }

    return counter;
  }
}
